spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/partitioner/ReplicaPartitioner.scala

```diff
@@ -0,0 +1,51 @@
+package com.datastax.spark.connector.rdd.partitioner
+
+import java.net.InetAddress
+
+import com.datastax.spark.connector.cql.CassandraConnector
+import org.apache.spark.{Partition, Partitioner}
+
+
+case class ReplicaPartition(index: Int, endpoints: Set[InetAddress]) extends EndpointPartition
+
+/**
+ * The replica partitioner will work on an RDD which is keyed on sets of InetAdresses representing Cassandra
+ * Hosts . It will group keys which share a common IP address into partitionsPerReplicaSet Partitions.
+ * @param partitionsPerReplicaSet The number of Spark Partitions to make Per Unique Endpoint
+ * @param connector
+ */
+class ReplicaPartitioner(partitionsPerReplicaSet: Int, connector: CassandraConnector) extends Partitioner {
```

Comment:
I think what I can do is create another RDD type which just defines the getPreferredLocation/partition method and have the final action of this code to be transform to that RDD. There will still be an additional move (from after the shuffle) but any RDD which references this one should pick up the correct location...